docs/lang/articles/misc/ndarray_android.md

```diff
@@ -0,0 +1,449 @@
+- With Ndarray, all you need to do is to update your runtime program.
+
+## Run a Taichi program using Ndarray on Android
+
+The following section walks you through the process of running a Taichi program using Ndarray on Android.
+
+1. [Generate necessary compute shaders](#generate-necessary-compute-shaders)
+2. [Parse the generated JSON file](#parse-the-generated-json-file)
+3. [Prepare SSBO and shape information](#prepare-ssbo-and-shape-information)
+4. [Prepare rendering shaders](#prepare-rendering-shaders)
+5. [Execute all shaders](#execute-all-shaders)
+
+:::note
+
+From Step 2, you are required to come up with your own runtime program. We provide an [example Java runtime program for Android](https://github.com/taichi-dev/taichi-aot-demo/blob/master/nbody_ndarray/java_runtime/NbodyNdarray.java) for your reference, but you may need to adapt these codes for your platform and programming language.
+
+:::
+
+### Generate necessary compute shaders
+
+The following Python script defines a Taichi AOT module for generating and saving the necessary compute shaders (GLES shaders in this case) based on the chosen backend (OpenGL).
+
+
+
+```Python
+import taichi as ti
+
+ti.init(arch=ti.opengl, use_gles=True, allow_nv_shader_extension=False)
+
+# Define constants for computation
+G = 1
+PI = 3.141592653
+N = 1000
+m = 5
+galaxy_size = 0.4
+planet_radius = 1
+init_vel = 120
+h = 1e-5
+substepping = 10
+
+# Define Taichi kernels
+@ti.kernel
+def initialize(pos: ti.any_arr(element_dim=1), vel: ti.any_arr(element_dim=1)):
+    center=ti.Vector([0.5, 0.5])
+    for i in pos:
+        theta = ti.random() * 2 * PI
+        r = (ti.sqrt(ti.random()) * 0.7 + 0.3) * galaxy_size
+        offset = r * ti.Vector([ti.cos(theta), ti.sin(theta)])
+        pos[i] = center+offset
+        vel[i] = [-offset.y, offset.x]
+        vel[i] *= init_vel
+
+@ti.kernel
+def compute_force(pos: ti.any_arr(element_dim=1), vel: ti.any_arr(element_dim=1), force: ti.any_arr(element_dim=1)):
+    for i in pos:
+        force[i] = ti.Vector([0.0, 0.0])
+    for i in pos:
+        p = pos[i]
+        for j in pos:
+            if i != j:
+                diff = p-pos[j]
+                r = diff.norm(1e-5)
+                f = -G * m * m * (1.0/r)**3 * diff
+                force[i] += f
+    dt = h/substepping
+    for i in pos:
+        vel[i].atomic_add(dt*force[i]/m)
+        pos[i].atomic_add(dt*vel[i])
+
+# Define Ndarrays
+pos = ti.Vector.ndarray(2, ti.f32, N)
+vel = ti.Vector.ndarray(2, ti.f32, N)
+force = ti.Vector.ndarray(2, ti.f32, N)
+
+# Run the AOT module builder
+def aot():
+    m = ti.aot.Module(ti.opengl)
+    m.add_kernel(initialize, (pos, vel))
+    m.add_kernel(compute_force, (pos, vel, force))
+
+    dir_name = 'nbody_aot'
+    m.save(dir_name, '')
+aot()
+```
+
+**In line 3, you initialize Taichi:**
+
+1. Set `use_gles` to `True` to generate GLES compute shaders on Android.
+2. Set `allow_nv_shader_extension` to `False` for the GLES compute shaders on Android.
```

Comment:
```suggestion
2. Set `allow_nv_shader_extension` to `False` to prevent the generated GLES compute shaders from using Nvidia GL extensions on Android.
```